---
title: "Data Science Project"
author: "Ethan Rosati"
date: "null"
output:
  html_document: default
  pdf_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Sequence Data

This [sequence data](https://github.com/rosatiem98/rosatiem98.github.io/blob/b683b09421b8428a629ebf75f9f32f0e1ba99ac7/plant_ITS%20(1).fasta) is a collection of DNA sequence of the entire ITS gene region in Angiosperm and Gymnosperm samples pulled from online databases. This set of [data](https://github.com/rosatiem98/rosatiem98.github.io/blob/c03555493c164c0240d596f179b77a41cbd2ff07/Research%20Sequences.docx) is a collection of DNA sequences gathered through DNA extraction and amplification of the ITS gene region in herbarium specimens. 

![Herbarium Specimen](./Gymnosperm specimen.jpg)

The purpose of this analysis is to determine the evolutionary differences between angiosperm and gymnosperm specimens. From analyzing differences a phylogenetic tree is built around the ITS gene region of these two clades. 

___
```{r echo=FALSE, message=FALSE, warning=FALSE}
library(tidyverse)
library(seqinr)
library(Biostrings)
library(msa)
library(phangorn)
library(ShortRead)
library(ape)
library(ggplot2)
library(ggtree)
library(ggmsa)
library(zoom)
```
___

Read in dataset from small research dataset FASTA file 
```{r message=FALSE, warning=FALSE}
example <- readDNAStringSet("../Data_Course_ROSATI/Final Project/Research sequences/DNA extraction sequences.fasta")

```
___

Alignment of research sequences
```{r message=FALSE, warning=FALSE}
#ggmsa alignment 

ggmsaresearch <- readRDS("../Data_Course_ROSATI/Final Project/ggmsa_resaerch_alignment.RDS")

ggmsaresearch

#MSA alignment
msaalign <- readRDS("../Data_Course_ROSATI/Final Project/research_sequences_alignment.RDS")
msaalign
print(msaalign, show="complete")
phang.alignex = as.phyDat(msaalign,type="DNA")

```
___

Conduct model test
```{r warning=FALSE, message=FALSE}
#researchmsa <- modelTest(phang.alignex)

mt <- readRDS("../Data_Course_ROSATI/Final Project/mtex_modeltest")
dmex <- dist.ml(phang.alignex, model="JC")
dmex <- readRDS("../Data_Course_ROSATI/Final Project/msaex_distance.RDS")
msaex_UPGMA <- upgma(dmex)
msaex_NJ <- NJ(dmex)
plot(msaex_UPGMA, main="UPGMA")

plot(msaex_NJ, main="NJ")

parsimony(msaex_UPGMA, phang.alignex)
parsimony(msaex_NJ, phang.alignex)


msaex_optim <- optim.parsimony(msaex_UPGMA, phang.alignex)

msaex_pratchet <- pratchet(phang.alignex)


plot(msaex_optim)

plot(msaex_pratchet)

```
___

Build a phylogenetic tree with bootsttrap values
```{r message=FALSE, warning=FALSE}
fitex = pml(msaex_UPGMA, data=phang.alignex)
print(fitex)

#fitJCex <- optim.pml(fitex,model="JC", rearrangement = "stochastic")

fitJCex <- readRDS("../Data_Course_ROSATI/Final Project/fitJCex.RDS")
logLik(fitJCex)

#bootstrapping

bsrex <- bootstrap.pml(fitJCex, bs=100, optNi=TRUE)
bsrex <- readRDS("../Data_Course_ROSATI/Final Project/boot_strp_research_seqs.RDS")
plotBS((fitJCex$tree), bsrex, p=50, type="p", bs.col = "red")
```



